### PR TITLE
MOVEONLY-ish: Do not include main.h from any other header

### PR DESCRIPTION
**Pull request by jtimon**
Moving some code, including main.h from qt/peertablemodel.h and wallet.h can be avoided.
Also, main.h doesn't need to include many things that can be included in main.cpp.
After that, using the compiler to make sure only files that depend on main.h include it, many missing includes show up.
Many files should require less memory for being compiled after this.

### CODE REVIEW

**Comment by theuni** [comment]
Very nice, I've been meaning to take a look at this as well. Great to see it was relatively non-invasive.
I've verified that fd10da8cae79a15d2ac23a17120cb2d62e8a28e2 and b46dd0f3f53ec20035660aaca07484f898df1fc5 are read-only. From a quick eye-ball (and since Travis is happy), the includes changes look good too.

utAck.

**Comment by laanwj** [comment]
ACK. Looks like a definite improvement to me, and low-risk due to move only.

**Comment by jtimon** [comment]
For b and c, maybe a half-solution that just comes to mind, is just leaving the definitions in main.cpp but expose them with an independent header like main_test.h or something. Just a random thought.

I'm basically focusing on consensus and policy checks. Trying to take advantage of potential similarities between their dependencies. Right now it's kind of a mess that doesn't even pass the tests, but if you're curious, here's my latest branch on that: https://github.com/jtimon/bitcoin/commits/consensus_policy
I guess it mostly falls into the "d" category, with a little bit of "a" as you move constants and globals out.
This (and therefore #5669 and the longest branch) needed rebase.
I'm also trying to isolate a good small base for policy to use as an example and confirm what I think @luke-jr and I (and hopefully everyone) can agree on.

**Comment by Diapolo** [comment]
Pretty cool, that you are following the ordering style still ^^. Does this also speedup compilation times, did you benchmark it?

**Comment by jtimon** [comment]
@Diapolo thanks for double-checking the include alphabetical order. 
I believe it does speedup compilation times, but I haven't benchmarked it. Some benchmarks would be great if you want to do them.

**Comment by jtimon** [comment]
I left the final commit cleaning up the includes for later.
It is the hardest to review. The remaining commits are practically move-only. The only change is creating ```CMainSignals& GetMainSignals()``` for accessing g_signals from main. So the PR becomes trivial to review and it's much less likely to need rebases.
On the other hand the cleaning up commit is likely to need frequent rebase and would likely break many open PRs so I've created #5697 which builds on top of this as well as on other useful and trivial preparations for consensus and policy code movements (#5696).

**Comment by jtimon** [comment]
Needed rebase.

**Comment by jtimon** [comment]
Needed rebase. I added a commit to not include main.h from any other header again, but it's not a big and exhaustive cleanup as it was at the beginning. I still leave that for #5697 after more code movements.

**Comment by theuni** [comment]
ut ACK, assuming travis is happy tomorrow. Verified that the MOVEONLY commits are still move only.

**Comment by theuni** [comment]
ACK. Could you please give this a quick bump for travis, though? There were some caching issues that have been long-fixed.

**Comment by jtimon** [comment]
Changed the last commit's id to give travis another try as demanded by @theuni

**Comment by theuni** [comment]
s/demanded by/humbly requested by/ :)

All green now, thanks.

**Comment by jtimon** [comment]
Sure, humbly requested, thanks for the review and ack.

**Comment by jtimon** [comment]
ping

**Comment by sipa** [comment]
The first commit doesn't make sense to me: CNodeStateStats is statistics of CNodeState, which is a main-specific data structure. Moving it up to a higher layer (net) seems wrong.

**Comment by jtimon** [comment]
How is net.o is higher level than main.o? main.cpp depends on net and not the other way around...

**Comment by sipa** [comment]
Maybe my terminology is confused. I mean main is a client of net, building
on top of it. Net should not know anything about what its clients do.
CNodeStateStats is stats about CNodeState, the message processing specific
data structure about nodes. Message processing is done in main, so this
data structure definition belongs there and nowhere else.

Don't get me wrong: the current placing of code in main and net is
terrible, but we shouldn't hack around that. Message processing needs to be
split up, and moved out to separate modules, and the related data
structures with it.

**Comment by jtimon** [comment]
Rebased without the first commit as suggested by @sipa

File: src/qt/signverifymessagedialog.cpp [hunk 1]
```diff
@@ -12,6 +12,7 @@
 
 #include "base58.h"
 #include "init.h"
+#include "main.h" // For strMessageMagic
```
Comment (sipa) [inline src/qt/signverifymessagedialog.cpp]: That's sad, we should move that elsewhere (not a criticism on your PR, just noticing).

**Comment by sipa** [comment]
Untested ACK.

**Comment by theuni** [comment]
utACK. I'd like to see the node info moved out as well, but that makes sense to do separately.

**Comment by jtimon** [comment]
Needed rebase, I also slightly improved the includes.
The commit that was criticized was already removed before needing the latest rebase.
This is quite uncontroversial, it has been reviewed, it has been open for long...is there any specific reason not to merge this?

**Comment by laanwj** [comment]
@jtimon No specific reason, just a overload of pulls caused me to lose track of this one for a while.

**Pull request merged**